### PR TITLE
[Merged by Bors] - feat(AlgebraicGeometry) define rational maps

### PR DESCRIPTION
**Pull request by erdOne**
and show that they are equivalent to `Spec X.functionField ⟶ Y`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/AlgebraicGeometry/RationalMap.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/RationalMap.lean:108]:
```suggestion
lemma fromSpecStalkOfMem_restrict (f : X.PartialMap Y)
    {U : X.Opens} (hU : Dense (U : Set X)) (hU' : U ≤ f.domain) {x} (hx : x ∈ U) :
    (f.restrict U hU hU').fromSpecStalkOfMem hx = f.fromSpecStalkOfMem (hU' hx) := by
```

File: Mathlib/AlgebraicGeometry/RationalMap.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/RationalMap.lean:130]:
```suggestion
`X` is irreducible and germ-injective at `x` (e.g. when `X` is integral),
```

File: Mathlib/AlgebraicGeometry/RationalMap.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/RationalMap.lean:138]: Instead of the `choose_spec.choose_spec.choose` stuff, could you add some API around `spread_out_of_isGermInjective` (i.e. add a new def using `choose` and the necessary lemmas extracted from `choose_spec`)?

File: Mathlib/AlgebraicGeometry/RationalMap.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/RationalMap.lean:173]:
```suggestion
lemma iseqv_rel : Equivalence (@Scheme.PartialMap.equiv X Y) where
```

File: Mathlib/AlgebraicGeometry/RationalMap.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/RationalMap.lean:219]:
```suggestion
  @Quotient (X.PartialMap Y) inferInstance
```
similarly below.

File: Mathlib/AlgebraicGeometry/RationalMap.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/RationalMap.lean:222]: Is this already typeable in VS code? (It is not in neovim)

File: Mathlib/AlgebraicGeometry/RationalMap.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/RationalMap.lean:240]: This docstring seems to be wrong.

File: Mathlib/AlgebraicGeometry/RationalMap.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/RationalMap.lean:253]: Ah it's just switched with this one.

File: Mathlib/AlgebraicGeometry/RationalMap.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/RationalMap.lean:289]: Same as in the module docstring.

File: Mathlib/AlgebraicGeometry/RationalMap.lean
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/RationalMap.lean:253]: Oops not sure how that happened.

File: Mathlib/AlgebraicGeometry/RationalMap.lean
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/RationalMap.lean:69]: Fair point. changed to `lemmasOnly`.

File: Mathlib/AlgebraicGeometry/RationalMap.lean
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/RationalMap.lean:138]: This def is intended to be the "new def using choose".

File: Mathlib/AlgebraicGeometry/RationalMap.lean
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/RationalMap.lean:222]: No. I will add a PR later.

File: Mathlib/AlgebraicGeometry/RationalMap.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/RationalMap.lean:134]:
```suggestion
noncomputable
def ofFromSpecStalk [PreirreducibleSpace X] [LocallyOfFiniteType sY] {x : X} [X.IsGermInjectiveAt x]
```
and below.

File: Mathlib/AlgebraicGeometry/RationalMap.lean
Comment (chrisflav) [reply Mathlib/AlgebraicGeometry/RationalMap.lean:138]: Okay, if we need it where `X` is not irreducible, we can still generalize.

File: Mathlib/AlgebraicGeometry/RationalMap.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/RationalMap.lean:71]: Maybe a lemma `restrict_restrict` would be useful?

File: Mathlib/AlgebraicGeometry/SpreadingOut.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/SpreadingOut.lean:353]:
```suggestion
Given `S`-schemes `X Y`, a point `x : X`, and a `S`-morphism `φ : Spec 𝒪_{X, x} ⟶ Y`,
we may spread it out to an `S`-morphism `f : U ⟶ Y`
```

File: Mathlib/AlgebraicGeometry/RationalMap.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/RationalMap.lean:294]: Lets not stall this PR, but I think that our current approach to treating `S`-schemes via carrying around the additional compatibility condition everywhere won't scale.

- Maybe we should migrate to an approach to AG where `X : Scheme` is replaced by `X : C` with `[HasForget₂ C Scheme]` with a coercion from `C` to `Scheme` (and for morphisms of `C`) whenever `HasForget₂ C Scheme`. This would give us more flexibility in working with `Over S`, because everything done for schemes still applies. The additional coercions everywhere might make this horrible though.

- Or we introduce `IsOver S` and `Hom.IsOver S`. Then objects are still schemes so no coercions, but we can comfortably have compositions be automatically `Hom.IsOver S` etc. This is also close to the rest of mathlibs design, where we replaced `Type` by `Scheme`.

Comment (erdOne) [reply Mathlib/AlgebraicGeometry/RationalMap.lean:294]: Yes this is on my radar as well. I'm thinking about the latter.

File: Mathlib/AlgebraicGeometry/RationalMap.lean
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/RationalMap.lean:134]: I don't think this matters because we assumed `X` is non-empty anyways (`by the `x : X`).

Comment (chrisflav) [reply Mathlib/AlgebraicGeometry/RationalMap.lean:149]: Oh yes, I did not pay attention.

File: Mathlib/AlgebraicGeometry/RationalMap.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/RationalMap.lean:208]:
```suggestion
```

File: Mathlib/AlgebraicGeometry/RationalMap.lean
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/RationalMap.lean:188]: This is a strange name, maybe just `equivalence`?

**Review by chrisflav — APPROVED** [review]
Thanks, LGTM modulo two minor comments.

**Comment by chrisflav** [comment]
Thanks!

maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**
